include/vigra/threadpool.hxx

```diff
@@ -0,0 +1,328 @@
+                        //
+                        // so the idea of this wait, is : If where are not in the destructor
+                        // (which sets stop to true, we wait here for new jobs)
+                        this->condition.wait(lock,[this]{ return this->stop || !this->tasks.empty(); });
+                        if(!this->tasks.empty()){
+                            ++busy;
+                            task = std::move(this->tasks.front());
+                            this->tasks.pop();
+                            lock.unlock();
+                            task(ti);
+                            ++processed;
+                            --busy;
+                            conditionF.notify_one();
+                        }
+                        else if(stop){
+                            return;
+                        }
+                    }
+                    
+                }
+            }
+        );
+    }
+}
+
+// add new work item to the pool
+template<class F>
+std::future<typename std::result_of<F(int)>::type> 
+ThreadPool::enqueueReturning(F&& f)
+{
+    typedef typename std::result_of<F(int)>::type result_type;
+    typedef std::future<result_type>  FutureResType;
+
+    auto task = std::make_shared< std::packaged_task<result_type(int)> >(
+        f                                                           
+    );
+
+    FutureResType res = task->get_future();
+    //std::future<return_type> res = task->get_future();
+    {
+        std::unique_lock<std::mutex> lock(queue_mutex);
+
+        // don't allow enqueueing after stopping the pool
+        if(stop)
+            throw std::runtime_error("enqueue on stopped ThreadPool");
+
+        tasks.emplace([task](int tid){ (*task)(tid); });
+    }
+    condition.notify_one();
+    return res;
+}
+
+// add new work item to the pool
+template<class F>
+void ThreadPool::enqueue(F&& f)
+{
+    //using return_type = typename std::result_of<F()>::type;
+    
+    //std::future<return_type> res = task->get_future();
+    {
+        std::unique_lock<std::mutex> lock(queue_mutex);
+
+        // don't allow enqueueing after stopping the pool
+        if(stop)
+            throw std::runtime_error("enqueue on stopped ThreadPool");
+
+        tasks.emplace(f);
+    }
+    condition.notify_one();
+    //return res;
+}
+
+// the destructor joins all threads
+inline ThreadPool::~ThreadPool()
+{
+    {
+        std::unique_lock<std::mutex> lock(queue_mutex);
+        stop = true;
+    }
+    condition.notify_all();
+    for(std::thread &worker: workers)
+        worker.join();
+}
+
+
+
+template<class ITER, class F>
+void parallel_foreach_impl(
+    ThreadPool & pool,
+    const uint64_t nItems,                   
+    ITER iter, 
+    ITER end, 
+    F && f,
+    std::random_access_iterator_tag
+){
+    // typedef typename std::iterator_traits<ITER>::reference ReferenceType;
+    uint64_t workload = std::distance(iter, end);
```

Comment:
...good point (but in this very casse std::distance is for free since it is a rand access iterator) but still....should be removed